File: src/test/java/pl/training/workshop/m3/s09_lsp/S09EquivalenceTest.java
package pl.training.workshop.m3.s09_lsp;

import java.util.Set;
import java.util.stream.Stream;

import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

import pl.training.workshop.support.Scene;

/** Kasa (na zwykłej sali) i raport (na sali archiwalnej) działają tak samo we wszystkich wariantach. */
final class S09EquivalenceTest {
    @TestFactory
    Stream<DynamicTest> sellingAndReportingBehaveTheSame() {
        return Scene.<Integer, String>variants()
                .variant("start", seat -> {
                    var hall = new pl.training.workshop.m3.s09_lsp.start.Hall(10);
                    hall.reserve(1);
                    return new pl.training.workshop.m3.s09_lsp.start.BoxOffice().sell(hall, seat) + " | "
                            + new pl.training.workshop.m3.s09_lsp.start.OccupancyReport().describe(
                            new pl.training.workshop.m3.s09_lsp.start.ReadOnlyHall(10, Set.of(1, 2, seat)));
                })
                .variant("step1", seat -> {
                    var hall = new pl.training.workshop.m3.s09_lsp.step1.Hall(10);
                    hall.reserve(1);
                    return new pl.training.workshop.m3.s09_lsp.step1.BoxOffice().sell(hall, seat) + " | "
                            + new pl.training.workshop.m3.s09_lsp.step1.OccupancyReport().describe(
                            new pl.training.workshop.m3.s09_lsp.step1.ReadOnlyHall(10, Set.of(1, 2, seat)));
                })
                .variant("step2", seat -> {
                    var hall = new pl.training.workshop.m3.s09_lsp.step2.Hall(10);
                    hall.reserve(1);
                    return new pl.training.workshop.m3.s09_lsp.step2.BoxOffice().sell(hall, seat) + " | "
                            + new pl.training.workshop.m3.s09_lsp.step2.OccupancyReport().describe(
                            new pl.training.workshop.m3.s09_lsp.step2.ReadOnlyHall(10, Set.of(1, 2, seat)));
                })
                .expect("miejsce 5", 5, "sprzedano miejsce 5, wolnych: 8 | zajete 3 z 10")
                .expect("ostatnie miejsce", 10, "sprzedano miejsce 10, wolnych: 8 | zajete 3 z 10")
                .tests();
    }
}

File: csharp/tests/Training.Workshop.Tests/M3/S09Lsp/S09EquivalenceTest.cs
namespace Training.Workshop.Tests.M3.S09Lsp;

/// <summary>Kasa (na zwykłej sali) i raport (na sali archiwalnej) działają tak samo we wszystkich wariantach.</summary>
public sealed class S09EquivalenceTest
{
    private static readonly Support.Scene<int, string> Scene = Support.Scene.Variants<int, string>()
        .Variant("start", seat =>
        {
            var hall = new Training.Workshop.M3.S09Lsp.Start.Hall(10);
            hall.Reserve(1);
            return new Training.Workshop.M3.S09Lsp.Start.BoxOffice().Sell(hall, seat) + " | "
                + new Training.Workshop.M3.S09Lsp.Start.OccupancyReport().Describe(
                    new Training.Workshop.M3.S09Lsp.Start.ReadOnlyHall(10, new HashSet<int> { 1, 2, seat }));
        })
        .Variant("step1", seat =>
        {
            var hall = new Training.Workshop.M3.S09Lsp.Step1.Hall(10);
            hall.Reserve(1);
            return new Training.Workshop.M3.S09Lsp.Step1.BoxOffice().Sell(hall, seat) + " | "
                + new Training.Workshop.M3.S09Lsp.Step1.OccupancyReport().Describe(
                    new Training.Workshop.M3.S09Lsp.Step1.ReadOnlyHall(10, new HashSet<int> { 1, 2, seat }));
        })
        .Variant("step2", seat =>
        {
            var hall = new Training.Workshop.M3.S09Lsp.Step2.Hall(10);
            hall.Reserve(1);
            return new Training.Workshop.M3.S09Lsp.Step2.BoxOffice().Sell(hall, seat) + " | "
                + new Training.Workshop.M3.S09Lsp.Step2.OccupancyReport().Describe(
                    new Training.Workshop.M3.S09Lsp.Step2.ReadOnlyHall(10, new HashSet<int> { 1, 2, seat }));
        })
        .Expect("miejsce 5", 5, "sprzedano miejsce 5, wolnych: 8 | zajete 3 z 10")
        .Expect("ostatnie miejsce", 10, "sprzedano miejsce 10, wolnych: 8 | zajete 3 z 10");

    public static TheoryData<string> Cases => Scene.Tests();

    [Theory]
    [MemberData(nameof(Cases))]
    public void SellingAndReportingBehaveTheSame(string test) => Scene.Run(test);
}
